File: untitled folder/untitled folder 2/flutter_secure_biometrics/android/src/main/kotlin/com/example/flutter_secure_biometrics/SecureBiometricManager.kt
package com.example.flutter_secure_biometrics

import android.content.Context
import android.hardware.camera2.CameraAccessException
import android.hardware.camera2.CameraManager
import android.os.Build
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import java.util.concurrent.Executor

class SecureBiometricManager {
    
    fun isAvailable(context: Context): Boolean {
        val biometricManager = BiometricManager.from(context)
        return when (biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG)) {
            BiometricManager.BIOMETRIC_SUCCESS -> true
            else -> false
        }
    }
    
    fun getAvailableBiometrics(context: Context): List<String> {
        val availableBiometrics = mutableListOf<String>()
        val biometricManager = BiometricManager.from(context)
        
        // Check fingerprint
        if (biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG) == BiometricManager.BIOMETRIC_SUCCESS) {
            availableBiometrics.add("fingerprint")
        }
        
        // Check face recognition (Android 10+)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            try {
                val cameraManager = context.getSystemService(Context.CAMERA_SERVICE) as CameraManager
                val cameraIds = cameraManager.cameraIdList
                if (cameraIds.isNotEmpty()) {
                    availableBiometrics.add("face")
                }
            } catch (e: CameraAccessException) {
                // Face recognition not available
            }
        }
        
        // Check for camera-based health biometrics
        if (hasCamera(context)) {
            availableBiometrics.add("heartRate")
            availableBiometrics.add("bloodOxygen")
        }
        
        return availableBiometrics
    }
    
    fun registerAppBiometric(
        context: Context,
        activity: FragmentActivity,
        type: String,
        appId: String,
        keyAlias: String,
        metadata: Map<String, Any>?
    ): Boolean {
        return try {
            // Create biometric prompt for registration
            val executor: Executor = ContextCompat.getMainExecutor(context)
            val biometricPrompt = BiometricPrompt(activity, executor, object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    super.onAuthenticationError(errorCode, errString)
                }
                
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    super.onAuthenticationSucceeded(result)
                    // Store app-specific biometric template
                    storeAppBiometricTemplate(context, appId, type, keyAlias)
                }
                
                override fun onAuthenticationFailed() {
                    super.onAuthenticationFailed()
                }
            })
            
            val promptInfo = BiometricPrompt.PromptInfo.Builder()
                .setTitle("Register ${type.capitalize()} for $appId")
                .setSubtitle("This will create an app-specific biometric profile")
                .setNegativeButtonText("Cancel")
                .build()
            
            biometricPrompt.authenticate(promptInfo)
            true
        } catch (e: Exception) {
            false
        }
    }
    
    fun authenticate(
        context: Context,
        activity: FragmentActivity,
        config: Map<String, Any>,
        appId: String,
        reason: String,
        keystoreManager: KeystoreManager,
        livenessDetector: LivenessDetector,
        spoofingDetector: SpoofingDetector,
        callback: (Map<String, Any>) -> Unit
    ) {
        val executor: Executor = ContextCompat.getMainExecutor(context)
        val biometricPrompt = BiometricPrompt(activity, executor, object : BiometricPrompt.AuthenticationCallback() {
            override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                super.onAuthenticationError(errorCode, errString)
                callback(mapOf(
                    "isAuthenticated" to false,
                    "error" to getErrorType(errorCode),
                    "errorMessage" to errString.toString(),
                    "timestamp" to System.currentTimeMillis()
                ))
            }
            
            override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                super.onAuthenticationSucceeded(result)
                
                // Perform additional security checks
                val enabledBiometrics = config["enabledBiometrics"] as? List<String> ?: listOf("fingerprint")
                val livenessDetection = config["livenessDetection"] as? Boolean ?: true
                val antiSpoofing = config["antiSpoofing"] as? Boolean ?: true
                
                var authenticationValid = true
                val usedBiometrics = mutableListOf<String>()
                
                // Validate each biometric type
                for (biometricType in enabledBiometrics) {
                    if (keystoreManager.isKeyRegistered(appId, biometricType)) {
                        usedBiometrics.add(biometricType)
                        
                        // Perform liveness detection
                        if (livenessDetection && !livenessDetector.performDetection(context, biometricType)) {
                            authenticationValid = false
                            break
                        }
                        
                        // Perform spoofing detection
                        if (antiSpoofing && spoofingDetector.detectSpoofing(context, biometricType)) {
                            authenticationValid = false
                            break
                        }
                    }
                }
                
                if (authenticationValid) {
                    callback(mapOf(
                        "isAuthenticated" to true,
                        "usedBiometrics" to usedBiometrics,
                        "trustScore" to mapOf(
                            "value" to 1.0,
                            "timestamp" to System.currentTimeMillis(),
                            "factors" to mapOf<String, Double>(),
                            "reason" to "Successful biometric authentication"
                        ),
                        "timestamp" to System.currentTimeMillis()
                    ))
                } else {
                    callback(mapOf(
                        "isAuthenticated" to false,
                        "error" to "livenessDetectionFailed",
                        "errorMessage" to "Security validation failed",
                        "timestamp" to System.currentTimeMillis()
                    ))
                }
            }
            
            override fun onAuthenticationFailed() {
                super.onAuthenticationFailed()
                callback(mapOf(
                    "isAuthenticated" to false,
                    "error" to "authenticationFailed",
                    "errorMessage" to "Biometric authentication failed",
                    "timestamp" to System.currentTimeMillis()
                ))
            }
        })
        
        val promptInfo = BiometricPrompt.PromptInfo.Builder()
            .setTitle("Authenticate")
            .setSubtitle(reason)
            .setNegativeButtonText("Cancel")
            .build()
        
        biometricPrompt.authenticate(promptInfo)
    }
    
    fun getHealthBiometrics(context: Context): Map<String, Double> {
        val healthData = mutableMapOf<String, Double>()
        
        if (hasCamera(context)) {
            // Simulate heart rate measurement via camera
            healthData["heartRate"] = measureHeartRate(context)
            
            // Simulate blood oxygen measurement
            healthData["bloodOxygen"] = measureBloodOxygen(context)
        }
        
        return healthData
    }
    
    private fun hasCamera(context: Context): Boolean {
        return context.packageManager.hasSystemFeature(android.content.pm.PackageManager.FEATURE_CAMERA_ANY)
    }
    
    private fun measureHeartRate(context: Context): Double {
        // Placeholder for actual heart rate measurement using camera
        // In real implementation, this would analyze camera frames for pulse detection
        return 72.0 + (Math.random() * 20 - 10) // Simulate 62-82 BPM
    }
    
    private fun measureBloodOxygen(context: Context): Double {
        // Placeholder for actual SpO2 measurement using camera and flash
        // In real implementation, this would use red and infrared light analysis
        return 98.0 + (Math.random() * 4 - 2) // Simulate 96-100%
    }
    
    private fun storeAppBiometricTemplate(context: Context, appId: String, type: String, keyAlias: String) {
        // Store app-specific biometric template in encrypted storage
        val sharedPrefs = context.getSharedPreferences("secure_biometrics_$appId", Context.MODE_PRIVATE)
        sharedPrefs.edit()
            .putString("${type}_template", keyAlias)
            .putLong("${type}_registered", System.currentTimeMillis())
            .apply()
    }
    
    private fun getErrorType(errorCode: Int): String {
        return when (errorCode) {
            BiometricPrompt.ERROR_USER_CANCELED -> "userCancelled"
            BiometricPrompt.ERROR_NO_BIOMETRICS -> "noBiometricsEnrolled"
            BiometricPrompt.ERROR_HW_NOT_PRESENT -> "biometricNotAvailable"
            BiometricPrompt.ERROR_HW_UNAVAILABLE -> "hardwareError"
            BiometricPrompt.ERROR_LOCKOUT -> "tooManyAttempts"
            BiometricPrompt.ERROR_LOCKOUT_PERMANENT -> "tooManyAttempts"
            BiometricPrompt.ERROR_TIMEOUT -> "timeout"
            else -> "unknown"
        }
    }
}

File: untitled folder/untitled folder 2/flutter_secure_biometrics/ios/Classes/SecureBiometricManager.swift
import Foundation
import LocalAuthentication
import Security
import AVFoundation
import CoreMotion

class SecureBiometricManager {
    
    private let context = LAContext()
    
    func isAvailable() -> Bool {
        var error: NSError?
        return context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error)
    }
    
    func getAvailableBiometrics() -> [String] {
        var availableBiometrics: [String] = []
        
        var error: NSError?
        if context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) {
            switch context.biometryType {
            case .faceID:
                availableBiometrics.append("face")
            case .touchID:
                availableBiometrics.append("fingerprint")
            case .opticID:
                availableBiometrics.append("opticID")
            case .none:
                break
            @unknown default:
                break
            }
        }
        
        // Check for camera-based health biometrics
        if AVCaptureDevice.default(for: .video) != nil {
            availableBiometrics.append("heartRate")
            availableBiometrics.append("bloodOxygen")
        }
        
        // Check for voice capability
        if AVAudioSession.sharedInstance().isInputAvailable {
            availableBiometrics.append("voice")
        }
        
        return availableBiometrics
    }
    
    func registerAppBiometric(
        type: String,
        appId: String,
        metadata: [String: Any]?,
        keychainManager: KeychainManager,
        completion: @escaping (Bool, Error?) -> Void
    ) {
        let reason = "Register \(type.capitalized) for \(appId)"
        
        context.evaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, localizedReason: reason) { success, error in
            DispatchQueue.main.async {
                if success {
                    // Generate app-specific key in Secure Enclave
                    do {
                        let keyGenerated = try keychainManager.generateAppSpecificKey(appId: appId, biometricType: type)
                        if keyGenerated {
                            // Store registration metadata
                            self.storeAppBiometricMetadata(appId: appId, type: type, metadata: metadata)
                            completion(true, nil)
                        } else {
                            completion(false, NSError(domain: "SecureBiometrics", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to generate key"]))
                        }
                    } catch {
                        completion(false, error)
                    }
                } else {
                    completion(false, error)
                }
            }
        }
    }
    
    func authenticate(
        config: [String: Any],
        appId: String,
        reason: String,
        keychainManager: KeychainManager,
        livenessDetector: LivenessDetector,
        spoofingDetector: SpoofingDetector,
        completion: @escaping ([String: Any]) -> Void
    ) {
        let enabledBiometrics = config["enabledBiometrics"] as? [String] ?? ["fingerprint"]
        let livenessDetection = config["livenessDetection"] as? Bool ?? true
        let antiSpoofing = config["antiSpoofing"] as? Bool ?? true
        
        // Validate app-specific biometric registration
        var registeredBiometrics: [String] = []
        for biometricType in enabledBiometrics {
            if keychainManager.isKeyRegistered(appId: appId, biometricType: biometricType) {
                registeredBiometrics.append(biometricType)
            }
        }
        
        if registeredBiometrics.isEmpty {
            completion([
                "isAuthenticated": false,
                "error": "appBiometricNotRegistered",
                "errorMessage": "No app-specific biometrics registered",
                "timestamp": Date().timeIntervalSince1970 * 1000
            ])
            return
        }
        
        // Perform biometric authentication
        let authContext = LAContext()
        authContext.evaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, localizedReason: reason) { success, error in
            DispatchQueue.main.async {
                if success {
                    // Perform additional security checks
                    self.performSecurityValidation(
                        biometrics: registeredBiometrics,
                        livenessDetection: livenessDetection,
                        antiSpoofing: antiSpoofing,
                        livenessDetector: livenessDetector,
                        spoofingDetector: spoofingDetector
                    ) { validationResult in
                        if validationResult.isValid {
                            completion([
                                "isAuthenticated": true,
                                "usedBiometrics": registeredBiometrics,
                                "trustScore": [
                                    "value": 1.0,
                                    "timestamp": Date().timeIntervalSince1970 * 1000,
                                    "factors": [:],
                                    "reason": "Successful biometric authentication"
                                ],
                                "timestamp": Date().timeIntervalSince1970 * 1000
                            ])
                        } else {
                            completion([
                                "isAuthenticated": false,
                                "error": validationResult.errorType,
                                "errorMessage": validationResult.errorMessage,
                                "timestamp": Date().timeIntervalSince1970 * 1000
                            ])
                        }
                    }
                } else {
                    let errorType = self.mapLAError(error)
                    completion([
                        "isAuthenticated": false,
                        "error": errorType,
                        "errorMessage": error?.localizedDescription ?? "Authentication failed",
                        "timestamp": Date().timeIntervalSince1970 * 1000
                    ])
                }
            }
        }
    }
    
    func getHealthBiometrics(completion: @escaping ([String: Double]?, Error?) -> Void) {
        var healthData: [String: Double] = [:]
        
        // Simulate heart rate measurement via camera
        if AVCaptureDevice.default(for: .video) != nil {
            measureHeartRate { heartRate in
                healthData["heartRate"] = heartRate
                
                // Simulate blood oxygen measurement
                self.measureBloodOxygen { bloodOxygen in
                    healthData["bloodOxygen"] = bloodOxygen
                    completion(healthData, nil)
                }
            }
        } else {
            completion(nil, NSError(domain: "SecureBiometrics", code: -1, userInfo: [NSLocalizedDescriptionKey: "Camera not available"]))
        }
    }
    
    private func performSecurityValidation(
        biometrics: [String],
        livenessDetection: Bool,
        antiSpoofing: Bool,
        livenessDetector: LivenessDetector,
        spoofingDetector: SpoofingDetector,
        completion: @escaping (ValidationResult) -> Void
    ) {
        let group = DispatchGroup()
        var validationPassed = true
        var errorType = ""
        var errorMessage = ""
        
        for biometricType in biometrics {
            // Perform liveness detection
            if livenessDetection {
                group.enter()
                livenessDetector.performDetection(type: biometricType) { isLive, error in
                    if !isLive || error != nil {
                        validationPassed = false
                        errorType = "livenessDetectionFailed"
                        errorMessage = "Liveness detection failed"
                    }
                    group.leave()
                }
            }
            
            // Perform spoofing detection
            if antiSpoofing {
                let isSpoofing = spoofingDetector.detectSpoofing(type: biometricType)
                if isSpoofing {
                    validationPassed = false
                    errorType = "spoofingDetected"
                    errorMessage = "Spoofing attempt detected"
                }
            }
        }
        
        group.notify(queue: .main) {
            completion(ValidationResult(
                isValid: validationPassed,
                errorType: errorType,
                errorMessage: errorMessage
            ))
        }
    }
    
    private func storeAppBiometricMetadata(appId: String, type: String, metadata: [String: Any]?) {
        let key = "secure_biometric_\(appId)_\(type)_metadata"
        if let metadata = metadata {
            UserDefaults.standard.set(metadata, forKey: key)
        }
        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: "\(key)_timestamp")
    }
    
    private func measureHeartRate(completion: @escaping (Double) -> Void) {
        // Simulate heart rate measurement using camera
        // In real implementation, this would analyze camera frames for pulse detection
        DispatchQueue.global().asyncAfter(deadline: .now() + 1.0) {
            let heartRate = 72.0 + Double.random(in: -10...10) // Simulate 62-82 BPM
            DispatchQueue.main.async {
                completion(heartRate)
            }
        }
    }
    
    private func measureBloodOxygen(completion: @escaping (Double) -> Void) {
        // Simulate SpO2 measurement using camera and flash
        // In real implementation, this would use red and infrared light analysis
        DispatchQueue.global().asyncAfter(deadline: .now() + 1.5) {
            let bloodOxygen = 98.0 + Double.random(in: -2...2) // Simulate 96-100%
            DispatchQueue.main.async {
                completion(bloodOxygen)
            }
        }
    }
    
    private func mapLAError(_ error: Error?) -> String {
        guard let laError = error as? LAError else { return "unknown" }
        
        switch laError.code {
        case .userCancel:
            return "userCancelled"
        case .biometryNotEnrolled:
            return "noBiometricsEnrolled"
        case .biometryNotAvailable:
            return "biometricNotAvailable"
        case .biometryLockout:
            return "tooManyAttempts"
        case .authenticationFailed:
            return "authenticationFailed"
        case .systemCancel:
            return "userCancelled"
        case .passcodeNotSet:
            return "biometricNotAvailable"
        default:
            return "unknown"
        }
    }
}

struct ValidationResult {
    let isValid: Bool
    let errorType: String
    let errorMessage: String
}
